src/drawing/fill.rs

```diff
@@ -0,0 +1,62 @@
+use image::{GenericImage, Pixel};
+
+/// Equivalent to bucket tool in MS-PAINT
+/// Performs 4-way flood-fill based on this algorithm: https://en.wikipedia.org/wiki/Flood_fill#Span_filling
+pub fn flood_fill<I>(image: &mut I, x: u32, y: u32, fill_with: I::Pixel)
+where
+    I: GenericImage,
+{
+    let target = image.get_pixel(x, y);
+
+    let mut stack = Vec::new();
+
+    stack.push((x as i32, x as i32, y as i32, 1 as i32));
+    stack.push((x as i32, x as i32, y as i32 - 1, -1 as i32));
+
+    while !stack.is_empty() {
+        let (x1, x2, y, dy) = stack.pop().unwrap();
+        let mut x1 = x1;
+        let mut x = x1;
+        if inside(image, x, y, target) {
+            while inside(image, x - 1, y, target) {
+                image.put_pixel(x as u32 - 1, y as u32, fill_with);
+                x = x - 1;
+            }
+            if x < x1 {
+                stack.push((x, x1 - 1, y - dy, -dy))
+            }
+        }
+        while x1 <= x2 {
+            while inside(image, x1, y, target) {
+                image.put_pixel(x1 as u32, y as u32, fill_with);
+                x1 = x1 + 1;
+            }
+            if x1 > x {
+                stack.push((x, x1 - 1, y + dy, dy))
+            }
+            if x1 - 1 > x2 {
+                stack.push((x2 + 1, x1 - 1, y - dy, -dy))
+            }
+            x1 = x1 + 1;
+            while x1 < x2 && !inside(image, x1, y, target) {
+                x1 = x1 + 1
+            }
+            x = x1
```

Comment:
Is there a way to write this with less mutable state as that makes it very hard to read. Maybe using iterator over pixels with a filter maybe?

Reply:
I agree it is ugly but I'm not sure that's avoidable with this particular algorithm.
It is copied verbatim from pseudocode on the wikipedia page so at least there is some explanation there.
There are simpler algorithms that would be more readable but less performant. The most elegant one (recursive) causes a stack overflow.

Reply:
Ah ok, if it's copied from the wikipedia then I'm satisfied with the correctness.